### PR TITLE
Tenet Tracer

### PR DESCRIPTION
**Pull request by EtchProject**
### Which kind of PR do you create?

- [x] This PR introduces a new function/api for Qiling Framework.

This pull request adds a tracer for IDAPro's [Tenet](https://github.com/gaasedelen/tenet) plugin in the extensions. The code structure is based completely off of the coverage extension using the context manager, allowing other tracing formats to be easily added in the future.

trace

[Tenet trace format](https://github.com/gaasedelen/tenet/tree/master/tracers)

Currently supports:

- arm
- arm64 (Tenet only supports up to 32 registers, so only up to X29 could be added)
- x86
- x64

### Coding convention?

- [x] The new code conforms to Qiling Framework naming convention.
- [x] The imports are arranged properly.
- [x] Essential comments are added.
- [x] The reference of the new code is pointed out.

### Extra tests?

- [x]  No extra tests are needed for this PR.
- [ ] I have added enough tests for this PR.
- [ ] Tests will be added after some discussion and review.

### Changelog?

- [ ] This PR doesn't need to update Changelog.
- [x] Changelog will be updated after some proper review.
- [ ] Changelog has been updated in my PR.

### Target branch?

- [x] The target branch is dev branch.

### One last thing

- [x] I have read the [contribution guide](https://docs.qiling.io/en/latest/contribution/)

-----

### CODE REVIEW

**Comment by chinggg** [comment]
Thanks! It seems to be a different tracer compared with the existing trace extension made by @elicn

**Comment by EtchProject** [comment]
> Thanks! It seems to be a different tracer compared with the existing trace extension made by @elicn

No problem, figured I'd share the code I had to make anyways. Yeah that one is for viewing the straight disassembly trace out of the box. This one is loaded into IDA's Tenet so you can see how it ran through the decompiled binary, what basic blocks it took, etc..

**Comment by xwings** [comment]
Nice work! Welcome to Qiling!

**Review by xwings — APPROVED** [review]

**Review by kabeor — APPROVED** [review]

File: qiling/extensions/tracing/formats/base.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    """
+    An abstract base class for trace collectors.
+    To add support for a new coverage format, just derive from this class and implement
+    all the methods marked with the @abstractmethod decorator.
+    """
+    
+    def __init__(self):
+        super().__init__()
+
+    @property
```
Comment (elicn) [inline qiling/extensions/tracing/formats/base.py:19]: Aggregating `property` on top of `staticmethod` doesn't work as a static property anymore.
See: https://stackoverflow.com/questions/58871322/property-with-staticmethod-in-python-3 (see references on top)

File: qiling/extensions/tracing/formats/registers.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+
+
+class ArchRegs():
+    """
+    Stores architecture's registers relevant to tracing in a format
+    universally usable by different tracers
+    """
+
+    arm_registers = {
```
Comment (elicn) [inline qiling/extensions/tracing/formats/registers.py:12]: These registers mappings are [somewhat] a duplication of the arch classes registers mappings.
Can we re-use the existing ones to create these and avoid code duplication? (with some necessary modifications, of course)

File: qiling/extensions/tracing/formats/registers.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        '$r10': unicorn.x86_const.UC_X86_REG_R10,
+        '$r11': unicorn.x86_const.UC_X86_REG_R11,
+        '$r12': unicorn.x86_const.UC_X86_REG_R12,
+        '$r13': unicorn.x86_const.UC_X86_REG_R13,
+        '$r14': unicorn.x86_const.UC_X86_REG_R14,
+        '$r15': unicorn.x86_const.UC_X86_REG_R15,
+    }
+
+    def __init__(self, arch):
+        if isinstance(arch, arm.QlArchARM):
```
Comment (elicn) [inline qiling/extensions/tracing/formats/registers.py:98]: Wouldn't a dictionaty-based selection [by `arch.type`] look better then multiple `elif`s..?

File: qiling/extensions/tracing/formats/tenet.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
+# This code structure is copied and modified from the coverage extension
+
+import qiling
```
Comment (elicn) [inline qiling/extensions/tracing/formats/tenet.py:3]: Better do `from qiling import Qiling`.
If this is only for type annotations, then it's would be best to put it under `if TYPE_CHECKING` to reduce unnecesary runtime depedencies.

File: qiling/extensions/tracing/formats/tenet.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        self.register_values= dict()
+
+        # Initialize with ridiculous value so first delta isn't missed
+        for register in self.arch_regs.registers:
+            self.register_values[register] = 0xFEEDBABE
+
+
+    def _add_delta(self):
+        # Cover glitch cases where nothing changed
+        if self.current_delta != []:
```
Comment (elicn) [inline qiling/extensions/tracing/formats/tenet.py:34]: petty: Could omit the `!= []` part to make it look cleaner.

File: qiling/extensions/tracing/formats/tenet.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            # Join all delta fragments into delta line and append 
+            self.deltas.append(",".join(self.current_delta))
+            self.current_delta = []
+        return
+
+
+    @staticmethod
+    def mem_access_callback(ql, access, address, size, value, self):
```
Comment (elicn) [inline qiling/extensions/tracing/formats/tenet.py:42]: Type annotations (where possible) could be great to keep the code readable and maintainable.

File: qiling/extensions/tracing/formats/tenet.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        access_type = None
+        # Set delta based on access type
+        if access == UC_MEM_READ:
+            # Since we are reading memory, we just read it ourselves
+            access_type = "mr"
+            value = ql.mem.read(address, size)
+            delta = f"{access_type}={hex(address)}:{value.hex()}"
```
Comment (elicn) [inline qiling/extensions/tracing/formats/tenet.py:49]: petty: Use `f'{address:#x}'`  intead of `hex(address)`

File: qiling/extensions/tracing/formats/tenet.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        elif access == UC_MEM_WRITE:
+            # Hook is before it's written, so we have to use the "value"
+            access_type = "mw"
+            if ql.arch.endian == QL_ENDIAN.EL:
```
Comment (elicn) [inline qiling/extensions/tracing/formats/tenet.py:53]: petty: Use ternary operation when an `if`-`else` construct asigns a value to the same variable.

File: qiling/extensions/tracing/formats/tenet.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                endian = 'little'
+            else:
+                endian = 'big'
+            if value < 0:
```
Comment (elicn) [inline qiling/extensions/tracing/formats/tenet.py:57]: petty: Could be just `sign = (value < 0)`

File: qiling/extensions/tracing/formats/tenet.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    def deactivate(self):
+        self.ql.hook_del(self.code_callback)
+        self.ql.hook_del(self.mem_write_callback)
+        self.ql.hook_del(self.mem_read_callback)
+    
+    def dump_trace(self, trace_file: str):
+        with open(trace_file, "w") as trace:
+            # Write out each delta on a separate line
+            for delta in self.deltas:
```
Comment (elicn) [inline qiling/extensions/tracing/formats/tenet.py:102]: petty: Alternatively, `trace.write('\n'.join(self.deltas) + '\n')` could be faster and cleaner

**Review by elicn — COMMENTED** [review]
Thanks for the great contribution!
I haven't got the chance to review it on time, but here are my belated comments anyway - for your reference. Some are petty, but some are not (:

**Pull request merged**